oauth2/oauth2-jose/src/test/java/org/springframework/security/oauth2/jwt/NimbusJweEncoderTests.java

```diff
@@ -0,0 +1,484 @@
+							"Unable to convert '" + JoseHeaderNames.JKU + "' JOSE header to a URI", ex);
+				}
+			}
+
+			Map<String, Object> jwk = headers.getJwk();
+			if (!CollectionUtils.isEmpty(jwk)) {
+				try {
+					builder.jwk(JWK.parse(jwk));
+				}
+				catch (Exception ex) {
+					throw new IllegalArgumentException("Unable to convert '" + JoseHeaderNames.JWK + "' JOSE header",
+							ex);
+				}
+			}
+
+			String keyId = headers.getKeyId();
+			if (StringUtils.hasText(keyId)) {
+				builder.keyID(keyId);
+			}
+
+			URL x509Uri = headers.getX509Uri();
+			if (x509Uri != null) {
+				try {
+					builder.x509CertURL(x509Uri.toURI());
+				}
+				catch (Exception ex) {
+					throw new IllegalArgumentException(
+							"Unable to convert '" + JoseHeaderNames.X5U + "' JOSE header to a URI", ex);
+				}
+			}
+
+			List<String> x509CertificateChain = headers.getX509CertificateChain();
+			if (!CollectionUtils.isEmpty(x509CertificateChain)) {
+				builder.x509CertChain(x509CertificateChain.stream().map(Base64::new).collect(Collectors.toList()));
+			}
+
+			String x509SHA1Thumbprint = headers.getX509SHA1Thumbprint();
+			if (StringUtils.hasText(x509SHA1Thumbprint)) {
+				builder.x509CertThumbprint(new Base64URL(x509SHA1Thumbprint));
+			}
+
+			String x509SHA256Thumbprint = headers.getX509SHA256Thumbprint();
+			if (StringUtils.hasText(x509SHA256Thumbprint)) {
+				builder.x509CertSHA256Thumbprint(new Base64URL(x509SHA256Thumbprint));
+			}
+
+			String type = headers.getType();
+			if (StringUtils.hasText(type)) {
+				builder.type(new JOSEObjectType(type));
+			}
+
+			String contentType = headers.getContentType();
+			if (StringUtils.hasText(contentType)) {
+				builder.contentType(contentType);
+			}
+
+			Set<String> critical = headers.getCritical();
+			if (!CollectionUtils.isEmpty(critical)) {
+				builder.criticalParams(critical);
+			}
+
+			Map<String, Object> customHeaders = headers.getHeaders().entrySet().stream()
+					.filter((header) -> !JWEHeader.getRegisteredParameterNames().contains(header.getKey()))
+					.collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
+			if (!CollectionUtils.isEmpty(customHeaders)) {
+				builder.customParams(customHeaders);
+			}
+
+			return builder.build();
+		}
+
+	}
+
+	static class JoseToken extends AbstractOAuth2Token {
+
+		private final JoseHeader headers;
+
+		private final JosePayload<?> payload;
+
+		JoseToken(String tokenValue, Instant issuedAt, Instant expiresAt, JoseHeader headers, JosePayload<?> payload) {
+			super(tokenValue, issuedAt, expiresAt);
+			this.headers = headers;
+			this.payload = payload;
+		}
+
+		JoseHeader getHeaders() {
+			return this.headers;
+		}
+
+		JosePayload<?> getPayload() {
+			return this.payload;
+		}
+
+	}
+
+	static class JosePayload<T> {
+
+		private final T content;
+
+		JosePayload(T content) {
+			this.content = content;
+		}
+
+		T getContent() {
+			return this.content;
+		}
+
+	}
+
+	// @formatter:off
+	/*
+	 * IMPORTANT DESIGN DECISION
```

Comment:
@jzheaux FYI

Reply:
@jzheaux @jgrandja Does this still need attention?

Reply:
@rwinch I'm currently working on some updates that may very well resolve this. Let's keep it for now until I complete updates.